src/components/forms/Input.ts

```diff
@@ -0,0 +1,119 @@
+import {EventBus} from '../../modules/EventBus';
+import {Block} from '../../modules/Block';
+import {getValidationMessage} from './ValidationMessage';
+
+// имя поля для повторного ввода оканчивается на 2
+const repeatFieldNameSuffix = '2';
+
+type IncomingProps = {
+  name: string;
+  label?: string;
+  value?: string;
+  type?: string;
+  placeholder?: string;
+  error?: string;
+  readonly?: boolean;
+}
+
+type EventState = Record<string, Record<string, string>>;
+
+export class Input extends Block {
+  constructor(props: IncomingProps) {
+    super(props);
+    const self = this;
+    this.setProps({
+      value: props.value || props.placeholder || '',
+      onFocus: function(event: Event): void {
+        if (self.props.value !== '' &&
+             (!self.props.placeholder ||
+              self.props.placeholder !== self.props.value)) {
```

Comment:
такие сложные условия лучше разбивать на несколько простых и помещать их в boolean переменные

```suggestion
        const not = (condition: boolean) => !condition
        const empty = (str: string) => str === '' || str === null || str === undefined
         const same = (first: string, second: second) => first === second
        const { value, placeholder } = self.props
        
        if (not(same(placeholder, value))) {
```

Думаю, так более читаемо